Fixing issue #8770: Improved frequency parameter logic to set it to 'D' only if periods, start, or end are None.

rjavierch opened the pull request:
This Pull Request addresses the issue reported regarding the default setting of the frequency (freq) parameter in Xarray. Currently, the frequency is explicitly set to "D" by default, which may not always be appropriate behavior. The proposed improvement adjusts the logic to set the frequency to "D" only if any of the parameters (periods, start, or end) are None, providing more flexibility and aligning with expected behavior.

Changes:

Updated logic for setting the frequency parameter to consider periods, start, or end values.
Added checks to ensure that the frequency is set to "D" only when necessary.

Please review and provide feedback on the proposed changes. Any suggestions for further improvements are welcome.

<!-- Feel free to remove check-list items aren't relevant to your change -->

- [x] Closes #8770

mathause commented:
Looks good. Can you add a small test (let us know if you need support for that) and maybe an entry to the whats-new doc.

rjavierch commented:
> Looks good. Can you add a small test (let us know if you need support for that) and maybe an entry to the whats-new doc.

Oh, yes, I would like to receive the support for that. I would love to add a small test and an entry to the whats-new doc.

spencerkclark commented on xarray/coding/cftime_offsets.py:
```diff
@@ -1218,6 +1218,9 @@ def date_range(
     """
     from xarray.coding.times import _is_standard_calendar
 
+    if freq is None and any(arg is None for arg in [periods, start, end]):
+        freq = "D"
```
Sorry on second thought could you move this logic into `cftime_range` (defined a little farther up in this file)?  That way we can harmonize the default arguments across `pandas.date_range`, `xarray.cftime_range`, and `xarray.date_range`.

spencerkclark reviewed (commented):
Indeed thanks @rjavierch—just one minor rethink on my part.  In terms of testing, you could add a test similar to:
https://github.com/pydata/xarray/blob/8e1dfcf571e77701d5669fd31f1576aed1007b45/xarray/tests/test_cftime_offsets.py#L1730-L1735
Only in this case you would test the situation where `freq` is not provided as an argument to `date_range`, but `start`, `end`, and `periods` are.

rjavierch commented:
Thanks for the feedback! I have updated the `test_invalid_cftime_range_inputs` inputs and added the `periods` argument as input in tests `test_cftime_range_no_freq` and `test_date_range_no_freq`.

mathause reviewed (approved):
LGTM

rjavierch commented:
Perfect! I close the issue with your approval.

mathause commented:
I reopen the PR - we still have to merge it :slightly_smiling_face: we can do that soon, but I want to give others the chance to have a look if they want

mathause commented:
Failure is unrelated.

spencerkclark commented on doc/whats-new.rst:
```diff
@@ -34,7 +34,9 @@ Deprecations
 Bug fixes
 ~~~~~~~~~
-
+- The logic of the frequency parameter in :py:meth:`xr.date_range` and :py:meth:`xr.cftime_range` are
+  set to ``'D'`` only if periods, start, or end are ``None`` (:issue:`8770`, :pull:`8774`).
+  By `Roberto Chang <https://github.com/rjavierch>`_.
```
```suggestion
- The default ``freq`` parameter in :py:meth:`xr.date_range` and :py:meth:`xr.cftime_range` is
  set to ``'D'`` only if ``periods``, ``start``, or ``end`` are ``None`` (:issue:`8770`, :pull:`8774`).
  By `Roberto Chang <https://github.com/rjavierch>`_.
```

spencerkclark reviewed (approved):
Looks good to me too—thanks @rjavierch!

rjavierch commented:
> I reopen the PR - we still have to merge it 🙂 we can do that soon, but I want to give others the chance to have a look if they want

Oh true, my bad! if there is something extra I need to do, please do not hesitate to let me know.

spencerkclark commented:
All good @rjavierch—merging now.
